test_ros2cli/CMakeLists.txt

```diff
@@ -19,14 +19,35 @@ if(BUILD_TESTING)
   find_package(ros_testing REQUIRED)
   find_package(rmw_implementation_cmake REQUIRED)
 
-  macro(custom_test_process_output test_name config_file_name timeout delay)
+  macro(add_custom_cli_test test_name)
     # Set variables for configuring the test files.
-    set(TEST_NAME ${test_name})  # test name used in the test cases
-    set(TEST_CONFIGURATION_FILE ${config_file_name})  # config file to use
-    set(DELAY ${delay})  # Delay between the actions are launch and the ros2 command is run
+    cmake_parse_arguments(TEST "" "CONFIG_MODULE;SETUP_DELAY;TIMEOUT" "" ${ARGN})
+    # Name for test cases
+    set(TEST_NAME ${test_name})
+    # RMW implementation for test cases
+    set(TEST_RMW_IMPLEMENTATION ${rmw_implementation})
+    # Module carrying test configuration
+    if(NOT DEFINED TEST_CONFIG_MODULE)
+      set(TEST_CONFIG_MODULE ${test_name}_config)
+    endif()
+    # Setup delay for test fixture actions
+    if(NOT DEFINED TEST_SETUP_DELAY)
+      set(TEST_SETUP_DELAY 2.0)
+      if("${TEST_RMW_IMPLEMENTATION}" STREQUAL "rmw_connext_cpp")
+        # Connext startup is too slow. It needs a few extra seconds till
+        # discovery completes.
+        set(TEST_SETUP_DELAY 5.0)
```

Comment:
That looks problematic because it:
* is fragile in case Connext might just take a second longer
* is slowing down the tests in case Connext happens to be faster than the threshold
* needs to be done for potentially other RMW impl

Instead we should find a way to write the test to work without this delay.

Reply:
It is problematic, I know. Truth to be told, all these tests have a delay to deal with CPU load fluctuations, uneven discovery times, etc., which makes them all flakey. I had to remove publishing/echoing rate tests because it's completely inconsistent, time wise.

Now, to fix this. Many tests don't need this delay at all e.g. `ros2 msg list`, so we can cut down time there. For those that do need it, I see two kinds: those that are bound to communication latency and those that are bound to discovery time. 

For the first kind, we can wait for specific output or events e.g. for `ros2 topic bw` wait till we have _any_ (non filtered) output to start testing it. 

For the second kind e.g. `ros2 topic list`, well, it gets harder. Assuming the daemon is there, we may repeatedly execute the command, though that isn't really less fragile than a delay (and I can't figure how to pull that off using `ros_testing` as it is today). We may also wait for the ROS graph to show some nodes, but since the CLI command, the CLI daemon and launch test are three separate processes, it may not work as intended.

Alternatively, I think that we need to make our CLI more predictable. Just like `ros2 service call` awaits for the server to be available, `ros2 service list` could have an optional ROS graph condition to wait for e.g. `ros2 service list --when "'my_server' in nodes"`. The same goes for all discovery time bound commands. What do you think?

Reply:
> Assuming the daemon is there, we may repeatedly execute the command, though that isn't really less fragile than a delay.

Why would that not be less fragile? It will pass the test as soon as the command returns the desired output and only runs to a maximum timeout if the command doesn't do so.

> Just like `ros2 service call` awaits for the server to be available, `ros2 service list` could have an optional ROS graph condition to wait for e.g. `ros2 service list --when "'my_server' in nodes"`.

This is a good example how *not* to add conditions imo. If you want to wait for X (services of a node being available) you can't use a condition Y (check if a node has been discovered). Simply because there is a still a race since the node might become available first shortly before the services are actually advertised.

Reply:
> Why would that not be less fragile? It will pass the test as soon as the command returns the desired output and only runs to a maximum timeout if the command doesn't do so.

I meant neither option makes it not fragile. You may not delay enough or you may not wait enough. Sure, we can increase the timeout a lot to increase the likelihood of discovery to find all peers for free, but that doesn't make it not fragile. If CI nodes are running with enough load, tests may still flake.

Anyways, even if we go down that route, we lack support in launch testing to do it so we have to go back to it first.

> Simply because there is a still a race since the node might become available first shortly before the services are actually advertised.

Yes, it doesn't make it not fragile either. I should've said that. I'm not well versed enough on DDS discovery mechanisms, is it atomic in any aspect e.g. some information chunks that are guaranteed to arrive in single block?

Reply:
> Sure, we can increase the timeout a lot to increase the likelihood of discovery to find all peers for free, but that doesn't make it not fragile. If CI nodes are running with enough load, tests may still flake.

I don't think that is the definition of flaky. The fact that we have to define an upper bound how long we are willing to wait is fine.